core/dbt/contracts/graph/model_config.py

```diff
@@ -1,87 +1,23 @@
 from dataclasses import field, Field, dataclass
-from enum import Enum
-from itertools import chain
-from typing import Any, List, Optional, Dict, Union, Type, TypeVar, Callable
+from typing import Any, List, Optional, Dict, Union, Type
 from typing_extensions import Annotated
 
+from dbt.common.contracts.config.base import BaseConfig, MergeBehavior, CompareBehavior
+from dbt.common.contracts.config.metadata import Metadata
 from dbt.common.dataclass_schema import (
     dbtClassMixin,
     ValidationError,
     StrEnum,
 )
-from dbt.contracts.graph.unparsed import AdditionalPropertiesAllowed, Docs
+from dbt.contracts.graph.unparsed import Docs
 from dbt.contracts.graph.utils import validate_color
 from dbt.contracts.util import Replaceable, list_str
-from dbt.common.exceptions import DbtInternalError, CompilationError
 from dbt import hooks
 from dbt.node_types import NodeType, AccessType
 from dbt_semantic_interfaces.type_enums.export_destination_type import ExportDestinationType
 from mashumaro.jsonschema.annotations import Pattern
 
 
-M = TypeVar("M", bound="Metadata")
-
-
-def _get_meta_value(cls: Type[M], fld: Field, key: str, default: Any) -> M:
-    # a metadata field might exist. If it does, it might have a matching key.
-    # If it has both, make sure the value is valid and return it. If it
-    # doesn't, return the default.
-    if fld.metadata:
-        value = fld.metadata.get(key, default)
-    else:
-        value = default
-
-    try:
-        return cls(value)
-    except ValueError as exc:
-        raise DbtInternalError(f"Invalid {cls} value: {value}") from exc
-
-
-def _set_meta_value(obj: M, key: str, existing: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
-    if existing is None:
-        result = {}
-    else:
-        result = existing.copy()
-    result.update({key: obj})
-    return result
-
-
-class Metadata(Enum):
-    @classmethod
-    def from_field(cls: Type[M], fld: Field) -> M:
-        default = cls.default_field()
-        key = cls.metadata_key()
-
-        return _get_meta_value(cls, fld, key, default)
-
-    def meta(self, existing: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
-        key = self.metadata_key()
-        return _set_meta_value(self, key, existing)
-
-    @classmethod
-    def default_field(cls) -> "Metadata":
-        raise NotImplementedError("Not implemented")
-
-    @classmethod
-    def metadata_key(cls) -> str:
-        raise NotImplementedError("Not implemented")
-
-
-class MergeBehavior(Metadata):
-    Append = 1
-    Update = 2
-    Clobber = 3
-    DictKeyAppend = 4
-
-    @classmethod
-    def default_field(cls) -> "MergeBehavior":
-        return cls.Clobber
-
-    @classmethod
-    def metadata_key(cls) -> str:
-        return "merge"
-
-
 class ShowBehavior(Metadata):
```

Comment:
I wonder if it's worth moving this class as well for completeness, even if it's not directly used in adapters. What do you think?